src/widgets/alsa_sound.rs

```diff
@@ -0,0 +1,138 @@
+                Err(::std::io::Error::new(kind, desc))
+            }
+        }
+    }
+    pub fn get_master_volume(&self) -> ::std::io::Result<i64> {
+        self.get_master()
+            .and_then(|master| {
+                master.get_playback_volume(SelemChannelId::mono())
+                      .map_err(|e| alsa_error_to_io_error("Failed to get `Master` volume", &e))
+            })
+    }
+    pub fn set_master_volume(&self, volume: i64) -> ::std::io::Result<()> {
+        self.get_master()
+            .and_then(|master| {
+                master.set_playback_volume_all(volume)
+                      .map_err(|e| alsa_error_to_io_error("Failed to set `Master` volume", &e))
+            })
+    }
+}
+
+pub struct AlsaVolume {
+    inner: Arc<Mutex<AlsaMixer>>
+}
+
+impl AlsaVolume {
+    pub fn initialize() -> ::std::io::Result<Self> {
+        AlsaMixer::new()
+            .map(|mixer| Self{ inner: Arc::new(Mutex::new(mixer)) })
+    }
+    pub fn get_master_volume(&self) -> ::std::io::Result<i64> {
+        self.inner
+            .try_lock()
+            .map_err(|_| {
+                let kind = ::std::io::ErrorKind::Other;
+                let msg = "Failed to get `Master` volume: locking failed";
+                ::std::io::Error::new(kind, msg)
+            })
+            .and_then(|ref alsa_mixer| alsa_mixer.get_master_volume())
+    }
+    
+    pub fn set_master_volume(&self, volume: i64) -> ::std::io::Result<()> {
+        self.inner
+            .try_lock()
+            .map_err(|_| {
+                let kind = ::std::io::ErrorKind::Other;
+                let msg = "Failed to set `Master` volume: locking failed";
+                ::std::io::Error::new(kind, msg)
+            })
+            .and_then(|ref alsa_mixer| alsa_mixer.set_master_volume(volume))
+    }
+    pub fn new(font_size: f32, length: u32) -> ::std::io::Result<Box<BarWidget>> {
+        Self::initialize().map(|alsa| BarWidget::new_simple(font_size, length, Box::new(alsa)))
+    }
+}
+
+
+// Note: ALSA returns values from 0 to 65536
+impl BarWidgetImpl for AlsaVolume {
+    fn name(&self) -> &str {
+        "volume"
+    }
+    fn value(&self) -> f32 {
+        // TODO: use min/max instead of hardcoded constants
+        self.get_master_volume()
+            .map(|volume| volume as f32 / 65536.0f32)
+            .unwrap_or_else(|e| { eprintln!("{}", e); 0.0f32 })
+    }
+    fn color(&self) -> Color {
+        // TODO: Custom color from config
+        Color::new(1.0, 1.0, 1.0, 1.0)
+    }
+    fn inc(&mut self, diff: f32) {
+        let r = self.inner
+            .try_lock()
+            .map_err(|_| {
+                let kind = ::std::io::ErrorKind::Other;
+                let msg = "Failed to set `Master` volume: locking failed";
+                ::std::io::Error::new(kind, msg)
+            })
+            .and_then(|ref mixer| {
+                let volume = mixer.get_master_volume()
+                    .map_err(|e| eprintln!("{}", e))
```

Comment:
Don't do random prints. You should probably just ? get_master_volume.

Reply:
It is not possible, because function returns `f32`, not `std::io::Result<f32>. I can just `unwrap_or`, but silent errors are not good, panicking is not good either because user wouldn't like launcher to fail when some of its parts don't work.

Reply:
This closure returns a `Result`.

Furthermore, depending on the situation, failing as a result of misconfiguration is acceptable. A call to `get_master_volume` in `new` could also be used to probe availability, removing the widget if it won't work at all.

Reply:
Yes, it does, but `inc` returns `f32`. The fail in this case is not a result of misconfiguration, it is an IO error. Furthermore, if call to `get_master_volume` succeeds once, it, strictly speaking, is not guaranteed to succeed in the future. However, if it succeeds once, there is almost 100% chance it will succeed again, so `Err` branch won't be triggered.
Probing sounds nice, I'll add it, but I still see no reason to neither remove error info nor panic on error.

Reply:
I'm OK with this PR as is after the feature is renamed, but:

> Yes, it does, but inc returns f32

I was referring to returning a result from the closure, which you already did for set_master_volume, to reuse your existing error handling rather than just printing an error and defaulting to 0, especially seeing that it's unlikely that set_master_volume will succeed if get_master_volume had failed.

For the error handling, the errors are unlikely to occur if it works once, which is also why I think having non-panic error handling is redundant as is. unwrap/expect is easier to read than nested and_then.

I am, however, not very fond of unstructured log output, which is just noise. There's no log facility in place, so widget errors are either that the widget does not appear (failed probe), or that wldash panics (which notifies the user that something is broken). Printing unstructured data to stderr that no one will see (starting wldash from an interactive terminal to monitor output is not a normal use-case), is less helpful than a panic.